darwin: Fix crash in log handler when stopping event thread

benzea opened the pull request:
The darwin event thread (in contrast to other OS implementations) tries
to log to the context that created this. However, this context is only
guaranteed to be valid until the thread has started. It may be that
another context is the last one to be destroyed, causing the event
thread to log using an already destroyed context.

Fix this by setting ctx to NULL and explicitly logging to a NULL
context.

Fixes: #1108

tormodvolden commented:
Actually the "setting ctx to NULL" seems not fixing anything, since it is not referenced later. I guess it is just a signal to the human code reader? Or is it confusing?

I see further up in the function that ctx is passed to some hotplug registration functions. I guess it is fine, if that is just being passed as a context pointer argument to the darwin_devices_detached() callback, since the latter is not using this argument. But if it is like this it we could use NULL there also.

benzea commented:
> Actually the "setting ctx to NULL" seems not fixing anything, since it is not referenced later. I guess it is just a signal to the human code reader? Or is it confusing?

Yeah, the variable is unused after that. I just thought I would make it explicit that the pointer must not be dereferenced at that point.

Anyway, I just changed it to add a comment at the start of the function.

> I see further up in the function that ctx is passed to some hotplug registration functions. I guess it is fine, if that is just being passed as a context pointer argument to the darwin_devices_detached() callback, since the latter is not using this argument. But if it is like this it we could use NULL there also.

I changed it to NULL. I do not know what that parameter means though. That said, the ptr parameter to the handler functions is unused in all cases. :-)

tormodvolden commented:
Yes, it seems the ptr parameter was unused ever since the introduction of darwin_devices_attached() in 9b76386 and ctx was used by habit.

tormodvolden commented:
Merged. Thanks for fixing yet another crash!
